lib/one_time_password/validator.rb

```diff
@@ -1,41 +1,36 @@
 module OneTimePassword
   class Validator < Base
-    def initialize(code, generated_at)
+    def initialize(code, generated_at = nil)
       @code = code
       @generated_at = generated_at
     end
 
     def valid?
-      code.present? && !wrong_length? && !expired? && !incorrect?
+      code.present? && !wrong_length? && (generated_at && !expired?) && !incorrect?
     end
 
     def warning
       return "Enter a passcode" if code.blank?
       return "Enter a valid passcode containing #{LENGTH} digits" if wrong_length?
-      return "Your passcode has expired, request a new one" if expired?
-      "Enter a valid passcode" if incorrect?
+      return "Your passcode has expired, request a new one" if generated_at && expired?
+      return "Your passcode is not valid or has expired" if !generated_at
+      "Enter a valid passcode"
     end
 
     private
 
     attr_reader :code, :generated_at
 
     def wrong_length?
-      return @wrong_length if defined?(@wrong_length)
-
-      @wrong_length = code.gsub(/\D/, "").length != LENGTH
```

Comment:
I'm not sure what this was ever doing - ignoring non-digits when checking for length doesn't seem useful.
e.g. code "1234A6B" would have the correct length in this case, but would still not be a valid code (it still fails the `ROTP:TOTP#verify`)